src/warm-storage/service.ts

```diff
@@ -335,76 +335,50 @@ export class WarmStorageService {
    * @returns Array of enhanced data set information
    */
   async getClientDataSetsWithDetails(client: string, onlyManaged: boolean = false): Promise<EnhancedDataSetInfo[]> {
-    const dataSets = await this.getClientDataSets(client)
     const pdpVerifier = this._getPDPVerifier()
     const viewContract = this._getWarmStorageViewContract()
 
-    // Process all data sets in parallel
-    const enhancedDataSetsPromises = dataSets.map(async (dataSet) => {
+    // Query dataset IDs directly from the view contract
+    const ids: bigint[] = await viewContract.clientDataSets(client)
+    if (ids.length === 0) return []
+
+    // Enhance all in parallel using dataset IDs
+    const enhancedDataSetsPromises = ids.map(async (idBigInt) => {
+      const pdpVerifierDataSetId = Number(idBigInt)
       try {
-        // Get the actual PDPVerifier data set ID from the rail ID (using pdpRailId now)
-        const pdpVerifierDataSetId = Number(await viewContract.railToDataSet(dataSet.pdpRailId))
-
-        // Check if this is a valid rail (rails should be > 0)
-        // Note: dataSetId can be 0 (legitimate first data set), but pdpRailId should never be 0
-        if (dataSet.pdpRailId === 0) {
-          return onlyManaged
-            ? null // Will be filtered out
-            : {
-                ...dataSet,
-                pdpVerifierDataSetId: 0,
-                nextPieceId: 0,
-                currentPieceCount: 0,
-                isLive: false,
-                isManaged: false,
-                withCDN: dataSet.cdnRailId > 0, // CDN is enabled if cdnRailId is non-zero (should be more reliable than metadata)
-                metadata: Object.create(null) as Record<string, string>,
-              }
-        }
+        const base = await this.getDataSet(pdpVerifierDataSetId)
 
-        // Parallelize independent calls
         const [isLive, listenerResult, metadata] = await Promise.all([
           pdpVerifier.dataSetLive(pdpVerifierDataSetId),
           pdpVerifier.getDataSetListener(pdpVerifierDataSetId).catch(() => null),
           this.getDataSetMetadata(pdpVerifierDataSetId).catch(() => Object.create(null) as Record<string, string>),
         ])
 
-        // Check if this data set is managed by our Warm Storage contract
         const isManaged =
           listenerResult != null && listenerResult.toLowerCase() === this._warmStorageAddress.toLowerCase()
 
-        // Skip unmanaged data sets if onlyManaged is true
-        if (onlyManaged && !isManaged) {
-          return null // Will be filtered out
-        }
+        if (onlyManaged && !isManaged) return null
```

Comment:
this didn't need to change, just leave the original form (I'm generally not a fan single-line if's)